### PR TITLE
Add "input" option to NumberType

### PR DESCRIPTION
**Pull request by webmozart**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24793
| License       | MIT
| Doc PR        | TODO

This PR replaces #24793 in (partially) fixing how Doctrine's DECIMAL type is handled by the Form component.

Previously, DECIMAL was mapped to the regular NumberType. That confuses Doctrine's change detection, depending on the DB platform.

Examples:

| DB | DB value | Doctrine entity before submit | Form input | Doctrine entity after submit
| --- | --- | --- | --- | ---
| SQLite | 8.000 | '8' | 8 | 8
| SQLite | 8.123 | '8.123' | 8.123 | 8.123
| PostgreSQL | 8.000 | '8.000' | 8 | 8
| PostgreSQL | 8.123 | '8.123' | 8.123 | 8.123

The value in the Doctrine entity changes before and after submit. Hence Doctrine believes an update is necessary.

This PR introduces an `input` option to NumberType (similar to DateType), that can be set to `'number'` (default) or `'string'`. If set to `'string'`, the conversion is as follows:

| DB | DB value | Doctrine entity before submit | Form input | Doctrine entity after submit
| --- | --- | --- | --- | ---
| SQLite | 8.000 | **'8'** | 8 | **'8.000'**
| SQLite | 8.123 | '8.123' | 8.123 | '8.123'
| PostgreSQL | 8.000 | '8.000' | 8 | '8.000'
| PostgreSQL | 8.123 | '8.123' | 8.123 | '8.123'

You see that this does not completely solve this issue for SQLite. However, @Ocramius and I agree that this is something to be fixed by Doctrine, since Doctrine is providing the database abstraction.

That fix should be done in the SqlitePlatform object in Doctrine as part of another PR and should be backwards compatible with the current Doctrine version (i.e. opt in via configuration).

Compared to #24793, this PR does not introduce a new type but instead makes the NumberType more flexible. Also, this PR does not introduce the `force_full_scale` option since that should be solved by Doctrine as described above.

### CODE REVIEW

**Comment by chalasr** [comment]
Can you add a note in the Bridge CHANGELOG?

**Comment by webmozart** [comment]
@chalasr Thanks for the heads up, done.

**Comment by xabbuh** [comment]
documentation PR added in symfony/symfony-docs#11314

File: src/Symfony/Bridge/Doctrine/CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ CHANGELOG
    an instance of `EntityManagerInterface` should be injected instead
  * added support for `simple_array` type
  * the `DoctrineTransactionMiddlewareFactory` class has been removed
+ * changed guessing of DECIMAL to set the `input` option of `NumberType` to string
```
Comment (fabpot) [inline src/Symfony/Bridge/Doctrine/CHANGELOG.md]: Should be in a  new section for 4.3 instead.

File: src/Symfony/Component/Form/Extension/Core/Type/NumberType.php [hunk 2]
```diff
@@ -33,6 +34,12 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             $options['rounding_mode'],
             $options['html5'] ? 'en' : null
         ));
+
+        if ('string' === $options['input']) {
+            $builder->addModelTransformer(new StringToFloatTransformer(
+                $options['scale']
+            ));
```
Comment (fabpot) [inline src/Symfony/Component/Form/Extension/Core/Type/NumberType.php]: should be on one line.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+class StringToFloatTransformer implements DataTransformerInterface
+{
+    /**
+     * @var int|null
+     */
+    private $scale;
+
+    /**
+     * @param int|null $scale
+     */
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php]: can be removed, no value according to typehinted arg, same for the property above.

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    public function __construct(int $scale = null)
+    {
+        $this->scale = $scale;
+    }
+
+    /**
+     * @param mixed $value
+     *
+     * @return string|null
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php]: `float|null`

**Review by yceruto — APPROVED** [review]

**Comment by yceruto** [comment]
Well, I guess that duplicated reviews will happen all the time, I'm sorry :)

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @webmozart.

**Pull request merged**